runtime/doc/lua.txt

```diff
@@ -1148,20 +1148,21 @@ vim.bo[{bufnr}]                                                                *
         print(vim.bo.comments)
         print(vim.bo.baz)                 -- error: invalid key
 <
-vim.wo[{winid}]                                                                *vim.wo*
-    Get or set window-scoped |options| for the window with handle {winid}.
-    Like `:set`. If [{winid}] is omitted then the current window is used.
-    Invalid {winid} or key is an error.
+vim.wo[{winid}][{bufnr}]                                                       *vim.wo*
```

Comment:
this seems like kind of a special case. 

or at least the "current buffer" case `vim.wo[foo][0]` would be addressed if `vim.lo.foo` or `vim.o.foo` worked like vim's `&foo`, right?

Reply:
No, because they set variables, not options.

This is supposed to be a replacement for `setlocal winopt`, or `nvim_[gs]et_option_value(..., {scope = 'local', win = winid, buf = bufnr })`.

But this makes it clearer that window local options are in fact specific to a buffer which current vimscript methods do not indicate at all.

Reply:
> No, because they set variables, not options.

`setwinvar()` / `setbufvar()` work with options too.




> But this makes it clearer that window local options are in fact specific to a buffer which current vimscript methods do not indicate at all.

that's a good argument in favor

> If you think it's esoteric, that's only because window options are an absolute shit show.

they are, which is why i was wondering if there's much need to make things ergonomic or Lua-land here: because options are already so complicated, if it's already possible to use vim functions to do what's needed, better to just link to those (and maybe provide examples if needed) rather than add an interface.

Reply:
Having spent _a lot_ of time thinking about options, I agree with Lewis here. We basically have two choices:

1. don't offer a Lua API for accessing options at all and funnel everybody through `nvim_{g,s}et_option_value`
2. make a **complete** and ergonomic Lua API from scratch.

1. is possible but would immediately lead to an explosion of a) complaints and b) user plugins offering a myriad of different Lua APIs, which would add horrible friction to the ecosystem. So that's out for me.

(Option 3., freezing/pruning the current API and leaving rough edges to be filled by Vimscript functions is also not really an option -- `setwinvar` works with options via `&option`, but the Lua API for variables does _not_, for example.)

This leaves option 2, where I believe that double-indexing is in fact the most explicit and ergonomic -- in fact, much better than Vimscript's -- way of dealing with window-local options (that are unfortunately not niche at all, since they include `number`).

Reply:
sounds like we have pretty strong grounding here

Reply:
Another thing to consider is that we could make all window options `global-local`, which is what I think most users expect how options to work anyway. i.e. `set number` in an `init.vim` enables line numbers by default, not, enables line numbers for the current window, from which new windows will inherit.

There would be a bunch of first and second order effects of doing that, but other than it being a (big) breaking change, I don't see any obvious negatives, it'll just be **a lot** of work.

One specific benefit that's relevant here is that, win-buf local options then become more of an advanced edge case, whereas now (as @clason says) are unfortunately not niche.